### PR TITLE
UI/notifications

### PR DESCRIPTION
**Pull request by nriedman**
# Build notification system

## :recycle: Current situation & Problem
Currently, the home view is empty. Moving forward, the UI for the Home Screen will include notifications, most recent vitals recorded, and a button to start the biweekly survey. To begin, we implemented a notification system that pulls notifications linked to the user's document in Firebase cloud storage, then displays them at the top of the dashboard in the Home Screen.

## :gear: Release Notes 
- Restructured the Home Screen dashboard as a sectioned list including Notifications
- Implemented a NotificationManager class that attaches a snapshot listener to a query on the currently signed in user's notification collection in Firestore
- The query retrieves notifications that are younger than 10 days old, and the snapshot listener in the NotificationManager then filters for notifications that haven't been completed yet.
- The NotificationManager is an environment accessible module that is configured in the ENGAGEDelegate.

## :books: Documentation
See inline documentation in NotificationManager.swift and Notification.swift.

## :white_check_mark: Testing
Thorough UI testing ensures that notifications are properly displayed when they are added to the user's backend Firestore, and that UI components such as the "mark complete" button and the "show more" / "show less" buttons work as intended.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [X] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PSchmiedmayer** [comment]
@nriedman Feel free to tag me here and drop me a Slack message once we have incorporated the elements we discussed about in the meeting yesterday 🚀

**Comment by nriedman** [comment]
@PSchmiedmayer I've gone ahead and incorporated the final changes we discussed. Ready for review!

File: ENGAGEHF/Dashboard/Notifications/NotificationManager.swift [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                    return nil
+                }
+                
+                self.logger.debug("Notifications updated")
+            }
+    }
+    
+    func add(_ notification: Notification) async {
+        await standard.add(notification: notification)
+    }
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/NotificationManager.swift:108]: I am wondering where this function is used and how it could be used in the future. If there is no specific use case I would suggest to remove it for now here and in the Standard.

File: ENGAGEHF/Dashboard/Notifications/NotificationManager.swift [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            return
+        }
+        
+        // Mark the notifications as completed in the Firestore
+        let timestamp = Timestamp(date: .now)
+        for offset in offsets {
+            let docID = notifications[offset].id
+            
+            let docRef = db.collection("users").document(user.uid).collection("notifications")
+                .document(docID)
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/NotificationManager.swift:131]: I would suggest a coherent formatting for elements like this. Either we have everything in a single line or spread across multiple lines.
```suggestion
            let docRef = db.collection("users")
                .document(user.uid)
                .collection("notifications")
                .document(docID)
```

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        
+        
+        var body: some View {
+            AsyncButton(
+                action: {
+                    let indx = notificationManager.notifications.firstIndex { $0.id == notification.id }
+                    
+                    if let indx: Int {
+                        await notificationManager.markComplete(at: IndexSet(integer: indx))
+                    }
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:57]: As we are matching ids to index paths and then back to ids in the notification manager I would suggest to transform this logic here and in the notification manager to actually work based on the notification id and not using the indirection using the index identification.

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    
+    
+    private struct XButton: View {
+        @Environment(NotificationManager.self) private var notificationManager
+        @Binding var notification: Notification
+        
+        
+        var body: some View {
+            AsyncButton(
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:51]: Nice use of the AsyncButton!

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+//
+// SPDX-License-Identifier: MIT
+//
+
+import SpeziViews
+import SwiftUI
+
+
+struct NotificationRow: View {
+    @Binding var notification: Notification
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:14]: I think that we can remove the `@Binding` here and in some similar views. There won't be any backward communication from here to the upper views and it seems like the notification can be a constant property as it will be never changed and modified which would need to be communicated up to parent views.

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:1]: Nice job with the layout here!

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            Text(notification.title)
+                .multilineTextAlignment(.leading)
+                .padding(.bottom, 10)
+            if isExpanded {
+                Text(notification.description)
+                    .multilineTextAlignment(.leading)
+                    .font(.footnote)
+            } else {
+                LearnMoreButton(isExpanded: $isExpanded)
+            }
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:37]: I am wondering if we should show at least one line and expand it to more lines (if available) using the learn more button. E.g. we can use the line limit modifier on a `Text` to set the line limit from 1 to 0 (I think that indicates that there is no line limit). So we have three states:

1. The Text fits in a single line (we could determine this like in this answer here: https://stackoverflow.com/a/75827200) and therefore there is no need for an expand button.
2. The Text doesn't fit in a single line and we are not in an expanded state: We show the text with a line limit of 1 and we show the show more button.
3. The Text doesn't fit in a single line and we are in an expanded state: We just show the text without any line limit.

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                        .foregroundStyle(.accent)
+                        .imageScale(.small)
+                }
+            )
+        }
+    }
+}
+
+
+#Preview {
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:70]: Nice job with the preview; great to see and makes it easy to review!

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    @Binding var notification: Notification
+    @State private var isExpanded = false
+    
+    
+    var body: some View {
+        VStack(alignment: .leading) {
+            HStack(alignment: .center) {
+                Text(notification.type)
+                    .font(.subheadline)
+                    .foregroundStyle(.secondary)
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:23]: Looking at the text in the Health App, we should make this a bit smaller, all uppercase, and a smaller padding to the top and bottom.

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                Spacer()
+                XButton(notification: $notification)
+            }
+            Divider()
+            Text(notification.title)
+                .multilineTextAlignment(.leading)
+                .padding(.bottom, 10)
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationRow.swift:30]: I would suggest that this text could be a bit smaller and could use a bold font to clearly highlight it as a title. The padding to the bottom could also be smaller.

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationSection.swift [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                    }
+                        .listRowSeparator(.hidden)
+                        .buttonStyle(.borderless)
+                } else {
+                    StudyApplicationListCard {
+                        HStack {
+                            Text("No new notifications")
+                            Spacer()
+                        }
+                    }
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationSection.swift:41]: Here is a nice [`ContentUnavailableView`](https://developer.apple.com/documentation/swiftui/contentunavailableview) in SwiftUI that you could use here, it has a nice standardized setup for views like this.

File: ENGAGEHF/Dashboard/Notifications/Views/NotificationSection.swift [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import SpeziViews
+import SwiftUI
+
+
+struct NotificationSection: View {
+    @Environment(NotificationManager.self) private var notificationManager
+    
+    
+    var body: some View {
+        @Bindable var notificationManager = notificationManager
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Notifications/Views/NotificationSection.swift:23]: If we don't make the notifications Bindings we could go ahead and remove the Bindable here.

File: ENGAGEHF/Dashboard/Survey/FullSurveyView.swift [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+// SPDX-License-Identifier: MIT
+//
+
+import SwiftUI
+
+
+struct FullSurveyView: View {
+    var body: some View {
+        Text("Fill in rest of survey here.")
+            .navigationTitle("KCCQ-12 Symptom Survey")
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Survey/FullSurveyView.swift:15]: We should either add the survey here using SpeziQuestionnaire or we should hide this view for now. Might be a separate PR and probably a lower-handing fruit for a small PR that we could do soon.

File: ENGAGEHF/Dashboard/Survey/SurveySection.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Survey/SurveySection.swift:1]: I am not 100% convinced about the UI design in this view. I would suggest to use a similar Card Layout that we use for the notifications section. It could inform about the Survey (maybe a few lines of Text) and the data it has been last filled out (and when it will be due again) or if it Is ready to be filled out we can use a larger primary bordered action button to indicate that one can do the survey.

File: ENGAGEHF/Dashboard/ToDo/Views/ToDoSection.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/Views/ToDoSection.swift:1]: From a UX perspective it might be nice to package this in a single card view that contains all TODOs (maybe with a divider between each them).

From a UI perspective: I think we could completely hide this section (not even a placeholder) if all TODOs are done to make the Home page a bit more decluttered.

File: ENGAGEHF/Dashboard/ToDo/Views/ToDoSection.swift [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+// SPDX-License-Identifier: MIT
+//
+
+import FirebaseAuth
+import FirebaseFirestore
+import SpeziViews
+import SwiftUI
+
+
+struct ToDoSection: View {
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/Views/ToDoSection.swift:15]: I am wondering if we should use the same manager setup as for the Notifications here as we do for the Notifications. It might be that we want to do that in a separate PR or include this in this PR.

File: ENGAGEHF/Dashboard/ToDo/Views/ToDoSection.swift [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    @State private var tasks: [ToDoItem] = []
+    @State private var state: ViewState = .idle
+    
+    @State private var tasksLoaded = false
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/Views/ToDoSection.swift:19]: It feels like `tasksLoaded` and `state` are displaying and contain the same state information. I think we could stick with the view state setup?

File: ENGAGEHF/Dashboard/ToDo/Views/ToDoRows.swift [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+import SwiftUI
+
+
+struct ToDoRow: View {
+    @Binding var task: ToDoItem
+    @State var taskComplete = false
+    
+    var body: some View {
+        HStack {
+            Text(task.task)
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/Views/ToDoRows.swift:18]: We might want to add a multiple text alignment here and in other sections to account for larger texts and text sizes.

File: ENGAGEHF/Dashboard/ToDo/Views/ToDoRows.swift [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+            Spacer()
+            Button {
+                taskComplete.toggle()
+            } label: {
+                Image(systemName: taskComplete ? "checkmark.square" : "square")
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/Views/ToDoRows.swift:23]: I think the task completion label could be a bit larger. You can use `.font(...)` on Images as well to make them scale with a dynamic text size and to make them, e.g., the same size as the text or proportionally larger.

File: ENGAGEHF/Dashboard/ToDo/Views/ToDoRows.swift [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+struct ToDoRows: View {
+    @Binding var tasks: [ToDoItem]
+    
+    var body: some View {
+        if !tasks.isEmpty {
+            ForEach($tasks, id: \.id) { task in
+                ToDoRow(task: task)
+            }
+        } else {
+            Text("Nothing to do!")
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/Views/ToDoRows.swift:40]: Similar to the Notifications we can either use off-the-shelf content unavailable views or rather hide the complete section if we don't have any TODOs.

We should identify a logic what TODOs should be there at the beginning when you open the app (we have that in the document) and what TODOs should reset on a daily basis (e.g. take a weight measurement).

File: ENGAGEHF/Dashboard/ToDo/ToDoItem.swift [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+
+
+struct ToDoItem: Identifiable {
+    let id: String
+    var task: String
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/ToDo/ToDoItem.swift:14]: I think we have a fixed set of TODOs and if they are only stored locally we could use an enum to e.g. define a clear set of TODOs and we rather display them (could have a computed property that presents a localized description).

File: ENGAGEHF/Dashboard/Vitals/Views/RecentVitalsSection.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Vitals/Views/RecentVitalsSection.swift:1]: As this view is very incomplete I am wondering if we should have it in this PR or rather make this a separate PR once we have all of the information in Firebase. We could start with Weight measurements in this PR?

We should use the weight measurement service (and this might be combined with the display of previous weight measurements).

I would suggest to display the following:
1. Value
2. When it was recorded
3. A nice visual icon about the type that was recorded

Tapping on them could navigate to the chart that we will implement in the future.

File: ENGAGEHF/ENGAGEHFStandard.swift [hunk 19]
```diff
@@ -63,19 +63,33 @@ actor ENGAGEHFStandard: Standard, EnvironmentAccessible, HealthKitConstraint, On
             _accountStorage = Dependency(wrappedValue: FirestoreAccountStorage(storeIn: ENGAGEHFStandard.userCollection))
         }
     }
+    
+    
+    // Adds a new notification to the user's notification collection
+    func add(notification: Notification) async {
```
Comment (PSchmiedmayer) [inline ENGAGEHF/ENGAGEHFStandard.swift:69]: As noted in the other review comment, it might not be necessary to add this function.

File: ENGAGEHF/Home.swift [hunk 20]
```diff
@@ -43,6 +47,21 @@ struct HomeView: View {
...
+            Medications(presentingAccount: $presentingAccount)
+                .tag(Tabs.medications)
+                .tabItem {
+                    Label("Medications", systemImage: "pill")
+                }
+            Education(presentingAccount: $presentingAccount)
+                .tag(Tabs.education)
+                .tabItem {
+                    Label("Education", systemImage: "brain")
+                }
```
Comment (PSchmiedmayer) [inline ENGAGEHF/Home.swift:64]: Wondering if we should already display them if they don't have any content as this point. We might add them as we add different content (e.g. a contacts section in the education tab and then include this in this PR).

File: ENGAGEHF/HeartHealth/HeartHealth.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/HeartHealth/HeartHealth.swift:1]: As noted in other comments it might make sense to keep this outside of the PR if we don't have a content to display here.

File: ENGAGEHF/Education/Education.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Education/Education.swift:1]: As noted in other comments it might make sense to keep this outside of the PR if we don't have a content to display here.

File: ENGAGEHF/Medications/Medications.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Medications/Medications.swift:1]: As noted in other comments it might make sense to keep this outside of the PR if we don't have a content to display here.

File: ENGAGEHF/ReusableElements/StudyApplicationListCard.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/ReusableElements/StudyApplicationListCard.swift:1]: We should add an issue in this repo to generalize this more and ideally move that (and even maybe the Notifications layout if we can make that generalizable including the text expansion functionality) into SpeziViews.

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for all the work in this PR @nriedman!
I fixed some very minor things in a small commit; the overall PR looks great!

It would be great if you can resolve all finished conversations and can create GitHub issues for all remaining elements that might still be open and should be addressed in future smaller or bigger PRs.
After this, feel free to merge the PR 🚀
